File: protocol/impls/java/Ecmh.java
import java.security.MessageDigest;

// ECMH state digest (§13.2) — the pinned, portable `sm ecmh` vector set.
//
// Mirrors impls/c/src/ecmh.c (ecmh_cmd) byte-for-byte: hash-to-curve KATs,
// accumulator algebra (identity / negate / add), and a tagged multiset sum,
// printed as a cross-language byte-identical `combined` golden. Runs Java's OWN
// secp256k1 (Secp.java) and must print output identical to the C reference.
final class Ecmh {

    private static MessageDigest sha256ctx() {
        try { return MessageDigest.getInstance("SHA-256"); }
        catch (Exception e) { throw new RuntimeException(e); }
    }

    // domain tags — second-preimage separation between tables.
    private static final byte TAG_NAME   = 0x01;
    private static final byte TAG_COMMIT = 0x02;
    private static final byte TAG_VOTE   = 0x03;
    private static final byte TAG_MUT    = 0x04;
    private static final byte[] ECMH_REC_TAG = { 'E','C','M','H','v','1' };

    // per-record H2C point: P("ECMHv1" ‖ tag ‖ body)
    private static byte[] recPt(byte tag, byte[] body) {
        byte[] pre = new byte[ECMH_REC_TAG.length + 1 + body.length];
        System.arraycopy(ECMH_REC_TAG, 0, pre, 0, ECMH_REC_TAG.length);
        pre[ECMH_REC_TAG.length] = tag;
        System.arraycopy(body, 0, pre, ECMH_REC_TAG.length + 1, body.length);
        return Secp.ecmhHash(pre)[0];
    }

    static void run() {
        StringBuilder out = new StringBuilder();
        MessageDigest comb = sha256ctx();

        // version self-doc
        out.append("ecmh ECMHv1\n"); comb.update(ECMH_REC_TAG);

        // 1. hash-to-curve KAT — fixed preimages → (ctr, compressed even-Y point).
        String[] labels = { "empty", "a", "shib", "doge", "ff32", "z32" };
        byte[][] pres = new byte[6][];
        pres[0] = new byte[0];
        pres[1] = Sm.utf8("a");
        pres[2] = Sm.utf8("shibpost");
        pres[3] = Sm.utf8("doge");
        pres[4] = new byte[32]; java.util.Arrays.fill(pres[4], (byte) 0xFF);
        pres[5] = new byte[32];
        for (int i = 0; i < 6; i++) {
            byte[][] h = Secp.ecmhHash(pres[i]);
            byte[] pt = h[0]; int ctr = h[1][0] & 0xFF;
            out.append("h2c ").append(labels[i]).append(" ctr=").append(ctr)
               .append(" pt=").append(Hex.enc(pt)).append("\n");
            comb.update((byte) ctr); comb.update(pt);
        }

        // 2. identity (∞) serialization
        byte[] id = Secp.ecmhIdentity();
        out.append("identity ").append(Hex.enc(id)).append("\n"); comb.update(id);

        // 3. tagged multiset sum — a fixed set of (tag ‖ body) records, summed two ways.
        byte[] tags = { TAG_NAME, TAG_NAME, TAG_COMMIT, TAG_VOTE, TAG_MUT };
        byte[][] bodies = {
            Sm.utf8("foo"),
            Sm.utf8("bar"),
            Sm.utf8("commitment-blob-32-bytes-xxxxxx"),
            Sm.utf8("vote-target-row"),
            Sm.utf8("owner-mutation"),
        };
        int nr = tags.length;
        byte[] fwd = Secp.ecmhIdentity(), rev = Secp.ecmhIdentity();
        for (int i = 0; i < nr; i++)        fwd = Secp.ecmhAdd(fwd, recPt(tags[i], bodies[i]));
        for (int i = nr - 1; i >= 0; i--)   rev = Secp.ecmhAdd(rev, recPt(tags[i], bodies[i]));
        int commut = java.util.Arrays.equals(fwd, rev) ? 1 : 0;
        out.append("sum ").append(Hex.enc(fwd)).append("\n");
        out.append("commutative ").append(commut).append("\n");
        comb.update(fwd); comb.update((byte) commut);

        // 4. inverse — remove the first record from the sum, re-add, must round-trip.
        byte[] pt0 = recPt(tags[0], bodies[0]);
        byte[] acc = fwd.clone();
        byte[] npt = pt0.clone(); Secp.ecmhNegate(npt);
        acc = Secp.ecmhAdd(acc, npt);   // remove rec[0]
        acc = Secp.ecmhAdd(acc, pt0);   // re-add rec[0]
        int roundtrip = java.util.Arrays.equals(acc, fwd) ? 1 : 0;
        out.append("inverse_roundtrip ").append(roundtrip).append("\n");
        comb.update((byte) roundtrip);

        byte[] cd = comb.digest();
        out.append("combined ").append(Hex.enc(cd)).append("\n");

        System.out.print(out);
    }
}

File: protocol/impls/csharp/Ecmh.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace Pepenet;

/// <summary>
/// §13.2 ECMH primitive vector set (`sm ecmh`). Mirrors impls/c/src/ecmh.c
/// (ecmh_cmd): hash-to-curve KATs, identity serialization, a tagged multiset sum
/// (commutativity + inverse round-trip), printed as a cross-language byte-identical
/// `combined` digest against this impl's own secp256k1. Names-only: no vote tag.
/// </summary>
public static class Ecmh
{
    // domain tags — second-preimage separation between tables (TAG_MUT keeps 0x04).
    private const byte TagName = 0x01, TagCommit = 0x02, TagMut = 0x04;
    private static readonly byte[] RecTag = { (byte)'E', (byte)'C', (byte)'M', (byte)'H', (byte)'v', (byte)'1' };

    public static int Run()
    {
        var sb = new StringBuilder();
        var comb = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        void Feed(byte[] d, int n) { comb.AppendData(d, 0, n); }
        void Feed1(byte b) => comb.AppendData(new[] { b }, 0, 1);

        // version self-doc
        sb.Append("ecmh ECMHv1\n"); Feed(RecTag, 6);

        // 1. hash-to-curve KAT — fixed preimages → (ctr, compressed even-Y point).
        var h2c = new (string label, byte[] pre)[]
        {
            ("empty", Array.Empty<byte>()),
            ("a",     Encoding.ASCII.GetBytes("a")),
            ("pepe",  Encoding.ASCII.GetBytes("pepenet")),
            ("doge",  Encoding.ASCII.GetBytes("doge")),
            ("ff32",  Fill(32, 0xFF)),
            ("z32",   new byte[32]),
        };
        foreach (var (label, pre) in h2c)
        {
            int ctr = Secp256k1.EcmhHash(pre, out byte[] pt);
            sb.Append("h2c ").Append(label).Append(" ctr=").Append(ctr).Append(" pt=").Append(Hx(pt)).Append('\n');
            Feed1((byte)ctr); Feed(pt, 33);
        }

        // 2. identity (∞) serialization
        byte[] id = Secp256k1.EcmhIdentity();
        sb.Append("identity ").Append(Hx(id)).Append('\n'); Feed(id, 33);

        // 3. tagged multiset sum — a fixed set of (tag ‖ row) records, summed two ways.
        var recs = new (byte tag, byte[] body)[]
        {
            (TagName,   Bytes(0x03, "foo")),
            (TagName,   Bytes(0x03, "bar")),
            (TagCommit, Encoding.ASCII.GetBytes("commitment-blob-32-bytes-xxxxxx")),
            (TagMut,    Encoding.ASCII.GetBytes("owner-mutation")),
        };
        byte[] fwd = Secp256k1.EcmhIdentity(), rev = Secp256k1.EcmhIdentity();
        for (int i = 0; i < recs.Length; i++)
        {
            Secp256k1.EcmhHash(RecPre(recs[i].tag, recs[i].body), out byte[] pt);
            Secp256k1.EcmhAdd(fwd, pt);
        }
        for (int i = recs.Length - 1; i >= 0; i--)
        {
            Secp256k1.EcmhHash(RecPre(recs[i].tag, recs[i].body), out byte[] pt);
            Secp256k1.EcmhAdd(rev, pt);
        }
        int commut = Eq(fwd, rev) ? 1 : 0;
        sb.Append("sum ").Append(Hx(fwd)).Append('\n');
        sb.Append("commutative ").Append(commut).Append('\n');
        Feed(fwd, 33); Feed1((byte)commut);

        // 4. inverse — remove the first record from the sum, re-add, must round-trip.
        {
            Secp256k1.EcmhHash(RecPre(recs[0].tag, recs[0].body), out byte[] pt0);
            byte[] acc = (byte[])fwd.Clone();
            byte[] npt = (byte[])pt0.Clone(); Secp256k1.EcmhNegate(npt);
            Secp256k1.EcmhAdd(acc, npt);   // remove rec[0]
            Secp256k1.EcmhAdd(acc, pt0);   // re-add rec[0]
            int roundtrip = Eq(acc, fwd) ? 1 : 0;
            sb.Append("inverse_roundtrip ").Append(roundtrip).Append('\n');
            Feed1((byte)roundtrip);
        }

        byte[] cd = comb.GetHashAndReset();
        sb.Append("combined ").Append(Hx(cd)).Append('\n');

        Console.Out.Write(sb.ToString());
        return 0;
    }

    private static byte[] RecPre(byte tag, byte[] body)
    {
        byte[] r = new byte[6 + 1 + body.Length];
        Buffer.BlockCopy(RecTag, 0, r, 0, 6);
        r[6] = tag;
        Buffer.BlockCopy(body, 0, r, 7, body.Length);
        return r;
    }

    private static byte[] Bytes(byte lead, string s)
    {
        byte[] t = Encoding.ASCII.GetBytes(s);
        byte[] r = new byte[1 + t.Length]; r[0] = lead;
        Buffer.BlockCopy(t, 0, r, 1, t.Length);
        return r;
    }
    private static byte[] Fill(int n, byte v) { byte[] r = new byte[n]; for (int i = 0; i < n; i++) r[i] = v; return r; }
    private static bool Eq(byte[] a, byte[] b)
    {
        if (a.Length != b.Length) return false;
        for (int i = 0; i < a.Length; i++) if (a[i] != b[i]) return false;
        return true;
    }
    private static string Hx(byte[] d) { var sb = new StringBuilder(d.Length * 2); foreach (var b in d) sb.Append(b.ToString("x2")); return sb.ToString(); }
}
